data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/SqsDynamicVisibilityTimeoutManager.java

```diff
@@ -0,0 +1,123 @@
+                initialMessageVisibilityTimeout - TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS);
+    }
+
+
+
+    public void startDynamicVisibilityForMessage(final Message message, final long objectSizeBytes) {
+        final int visibilityTimeoutIncrease = calculateVisibilityTimeoutFromObjectSize(objectSizeBytes);
+        final int waitBeforeIncreasingAgain = Math.min(MINIMUM_WAIT_BEFORE_CHANGE_VISIBILITY_TIMEOUT_SECONDS,
+                visibilityTimeoutIncrease - TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS);
+
+        final ScheduledFuture<?> scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(
+                () -> increaseVisibilityTimeout(visibilityTimeoutIncrease, message),
+                initialWaitBeforeIncrease,
+                waitBeforeIncreasingAgain, TimeUnit.SECONDS);
+
+        messageToScheduledFuture.put(message.messageId(), scheduledFuture);
+    }
+
+    public void stopDynamicVisibilityForMessage(final String messageId) {
+        final ScheduledFuture<?> scheduledFutureToCancel = messageToScheduledFuture.get(messageId);
+        if (Objects.nonNull(scheduledFutureToCancel)) {
+            boolean cancelled = scheduledFutureToCancel.cancel(true);
+
+            if (!cancelled) {
+                LOG.error("Unable to cancel dynamic visibility execution for message {}", messageId);
+            } else {
+                messageToScheduledFuture.remove(messageId);
+            }
+        }
+    }
+
+    public int getRemainingTimeBeforeNextChangeMessageIncrease(final String messageId) {
+        final ScheduledFuture<?> scheduledFutureToCancel = messageToScheduledFuture.get(messageId);
+        if (Objects.nonNull(scheduledFutureToCancel)) {
+            return (int) scheduledFutureToCancel.getDelay(TimeUnit.SECONDS);
+        }
+
+        return initialMessageVisibilityTimeout;
+    }
+
+    public void shutdownDynamicMessaging() {
+        try {
+            scheduledExecutorService.shutdown();
+            scheduledExecutorService.awaitTermination(3, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            LOG.error("Dynamic SQS messaging was interrupted during shutdown");
+        } finally {
+            if (!scheduledExecutorService.isTerminated()) {
+                scheduledExecutorService.shutdown();
+            }
+        }
+    }
+
+    private void increaseVisibilityTimeout(final int visibilityTimeoutIncrease,
+                                           final Message message) {
+
+        final ChangeMessageVisibilityRequest changeMessageVisibilityRequest = ChangeMessageVisibilityRequest.builder()
+                .receiptHandle(message.receiptHandle())
+                .visibilityTimeout(visibilityTimeoutIncrease)
+                .queueUrl(queueUrl)
+                .build();
+
+        try {
+            sqsClient.changeMessageVisibility(changeMessageVisibilityRequest);
```

Comment:
We should be performing this using [ChangeMessageVisibilityBatch](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ChangeMessageVisibilityBatch.html) for a couple reasons:

1. Doing this one-by-one will create quite a few threads.
2. This creates more HTTP requests than we need which uses resources unnecessarily.
3. SQS charges on a per HTTP request basis. We are making this more expensive for users.

Reply:
That's a lot of good reasons to switch to it, but this might be difficult since each message will need this method called at different intervals.